spring-integration-redis/src/main/java/org/springframework/integration/redis/store/RedisMessageStore.java

```diff
@@ -183,6 +211,83 @@ protected Collection<?> doListKeys(String keyPattern) {
 		return this.redisTemplate.keys(keyPattern);
 	}
 
+	@Override
+	protected MessageGroup copy(MessageGroup group) {
+		return lockExecute(group.getGroupId(), () -> super.copy(group));
+	}
+
+	@Override
+	public void addMessagesToGroup(Object groupId, Message<?>... messages) {
+		Assert.notNull(groupId, "'groupId' must not be null");
+
+		lockExecute(groupId, () -> {
+			super.addMessagesToGroup(groupId, messages);
+			return null;
+		});
+	}
+
+	@Override
+	public void removeMessageGroup(Object groupId) {
+		lockExecute(groupId, () -> {
+			super.removeMessageGroup(groupId);
+			return null;
+		});
+	}
+
+	@Override
+	public void removeMessagesFromGroup(Object groupId, Collection<Message<?>> messages) {
+		lockExecute(groupId, () -> {
+			super.removeMessagesFromGroup(groupId, messages);
+			return null;
+		});
+	}
+
+	@Override
+	public boolean removeMessageFromGroupById(Object groupId, UUID messageId) {
+		return lockExecute(groupId, () -> super.removeMessageFromGroupById(groupId, messageId));
+	}
+
+	@Override
+	public void setLastReleasedSequenceNumberForGroup(Object groupId, int sequenceNumber) {
+		lockExecute(groupId, () -> {
+			super.setLastReleasedSequenceNumberForGroup(groupId, sequenceNumber);
+			return null;
+		});
+	}
+
+	@Override
+	public void completeGroup(Object groupId) {
+		lockExecute(groupId, () -> {
+			super.completeGroup(groupId);
+			return null;
+		});
+	}
+
+	@Override
+	public void setGroupCondition(Object groupId, String condition) {
+		lockExecute(groupId, () -> {
+			super.setGroupCondition(groupId, condition);
+			return null;
+		});
+	}
+
+	public <T> T lockExecute(Object groupId, Supplier<T> supplier) {
+		Lock lock = this.lockRegistry.obtain(groupId);
+		try {
+			lock.lockInterruptibly();
+			try {
+				return supplier.get();
```

Comment:
Let me know, please, why `LockRegistry.executeLocked()` API doesn't work here?
I think you have extracted a common code from the `SimpleMessageStore`, but I guess the `LockRegistry.executeLocked()` could help us a lot. 

I don't say that `SimpleMessageStore` must not be fixed as well, but that is different story for different PR.